### PR TITLE
Allow connections from many clients simultaneously

### PR DESCRIPTION
**Pull request by masonforest**

### CODE REVIEW

File: app/controllers/api_tokens_controller.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
+class ApiTokensController < ApplicationController
+  prepend_before_filter :allow_params_authentication!, only: :create
+  api :POST, '/api_token', 'Creates a api_token'
```
Comment (masonforest) [inline app/controllers/api_tokens_controller.rb]: Looks like Hound updated its default double/single quote settings. I changed it back [here](https://github.com/projectjellyfish/api/commit/a794781c6a06f0eea394d278d340e9a00f4cf5fe).

**Comment by chriskacerguis** [comment]
:+1:

File: app/controllers/api_tokens_controller.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+  def create
```
Comment (calebthompson) [inline app/controllers/api_tokens_controller.rb]: apipie?

File: app/controllers/api_tokens_controller.rb [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    resource = warden.authenticate!(params[:staff])
+    api_token = resource.api_tokens.create
+
+    render json: { token: api_token.token }
+  end
+
+  def allow_params_authentication!
```
Comment (calebthompson) [inline app/controllers/api_tokens_controller.rb]: Should this be private? Public methods on controllers become routable (but not necessarily routed) actions.

File: app/controllers/concerns/token_authentication.rb [hunk 4]
```diff
@@ -7,36 +7,20 @@ module TokenAuthentication
...
-    user = user_email && Staff.find_by(email: user_email)
-
-    return unless user && user.token?(params[:user_token])
+    token = ApiToken.find_by_token(token_from_header || params[:access_token])
 
-    # Notice we are passing store false, so the user is not actually stored in the session and a token is needed for
-    # every request. If you want the token to work as a sign in token, you can simply remove store: false.
-    sign_in user, store: false
+    if token
+      sign_in(token.staff, store: false)
```
Comment (calebthompson) [inline app/controllers/concerns/token_authentication.rb]: Hmm, do we want tokens required on every request? I'm fine with it but it's going to complicate tests and such right?

Comment (calebthompson) [inline app/controllers/concerns/token_authentication.rb]: And does the front-end need to be updated for that change?

File: app/models/api_token.rb [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+#
+#  index_api_tokens_on_staff_id  (staff_id)
+#
+
+class ApiToken < ActiveRecord::Base
+  belongs_to :staff
+  before_create :generate_token
+
+  def generate_token
+    self.token = SecureRandom.hex(4)
```
Comment (calebthompson) [inline app/models/api_token.rb]: Let's make it longer than 4. Maybe 16? 32?

File: app/controllers/concerns/token_authentication.rb (hunk 4 above)
Comment (calebthompson) [inline app/controllers/concerns/token_authentication.rb]: Oh I guess this was already happening.

File: app/models/api_token.rb (hunk 5 above)
Comment (masonforest) [inline app/models/api_token.rb]: For some reason `SecureRandom.hex` generates a string that's twice as long as the length you pass in here.
`SecureRandom.hex(4).length # => 8`
Anyway I made it `16` (8) -> https://github.com/projectjellyfish/api/commit/223e0aae98afc5b79ab4ff7d97ec47a67ebe2488.

File: app/controllers/api_tokens_controller.rb (hunk 2 above)
Comment (masonforest) [inline app/controllers/api_tokens_controller.rb]: oh yep https://github.com/projectjellyfish/api/commit/1cf95f08bd893fb3ebd6e04dc6e19bdabd4a881f

File: app/controllers/api_tokens_controller.rb (hunk 3 above)
Comment (masonforest) [inline app/controllers/api_tokens_controller.rb]: good call https://github.com/projectjellyfish/api/commit/d61fc53551a9c6ea76556a6f1efd67f0d464f629

**Comment by chriskacerguis** [comment]
@masonforest looks like it failed a test, could you please take a look?

**Comment by masonforest** [comment]
Looks like changing the token length broke the tests. Fixed here -> https://travis-ci.org/projectjellyfish/api/builds/62279456

**Comment by chriskacerguis** [comment]
LGTM

**Comment by chriskacerguis** [comment]
:+1:

**Pull request merged**